### PR TITLE
Improve typing logic

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
This should make typing logic more stable and speed up several flows.

### CODE REVIEW

File: WebDriverAgentLib/Categories/XCUIElement+FBTyping.m [hunk 1]
```diff
@@ -13,15 +13,42 @@
 #import "FBKeyboard.h"
 #import "NSString+FBVisualLength.h"
 #import "XCUIElement+FBTap.h"
+#import "XCUIElement+FBUtilities.h"
 
 @implementation XCUIElement (FBTyping)
 
+- (BOOL)fb_prepareForTextInputWithError:(NSError **)error
+{
+  BOOL isKeyboardAlredyVisible = [FBKeyboard waitUntilVisibleForApplication:self.application timeout:-1 error:error];
```
Comment (imurchie) [inline WebDriverAgentLib/Categories/XCUIElement+FBTyping.m]: `isKeyboardAlreadyVisible`

File: WebDriverAgentLib/Categories/XCUIElement+FBTyping.m [hunk 2]
```diff
@@ -13,15 +13,42 @@
...
+  if (isKeyboardAlredyVisible && self.hasKeyboardFocus) {
+    return YES;
+  }
+  
+  int retriesCount = 2;
+  int tryNum = 0;
+  while (tryNum < retriesCount) {
```
Comment (imurchie) [inline WebDriverAgentLib/Categories/XCUIElement+FBTyping.m]: Does Objective-C not have a `for` loop that does this count/tries handling for you?

**Review by imurchie — COMMENTED** [review]
Nice!

Do none of the errors returned have any diagnostic value?

File: WebDriverAgentLib/Categories/XCUIElement+FBTyping.m (hunk 2 above)
Comment (mykola-mokhnach) [reply WebDriverAgentLib/Categories/XCUIElement+FBTyping.m]: what do you mean? any example in JS?

Comment (imurchie) [reply WebDriverAgentLib/Categories/XCUIElement+FBTyping.m]: Just a normal C-style `for` loop (`for (int i = 0; i < sizeof(something); i++) { ... }`)

File: WebDriverAgentLib/Categories/XCUIElement+FBTyping.m (hunk 1 above)
Comment (mykola-mokhnach) [reply WebDriverAgentLib/Categories/XCUIElement+FBTyping.m]: fixed

File: WebDriverAgentLib/Categories/XCUIElement+FBTyping.m (hunk 2 above)
Comment (mykola-mokhnach) [reply WebDriverAgentLib/Categories/XCUIElement+FBTyping.m]: changed to `for`

**Review by imurchie — APPROVED** [review]

**Pull request merged**